### ISSUE TITLE
feat(qwen3-tts): entropy-based degeneration guardrail for AR talker

### ISSUE THREAD

**Issue by marksverdhei** [heiervang-technologies/ht-vllm-omni#28]
## Problem

Qwen3-TTS occasionally enters a degeneration spiral during autoregressive codec token generation: initial stuttering (repeated N-grams) accumulates errors from previous bad tokens, which compounds into chaotic noise output. By the time `max_tokens` terminates the sequence, the audio is already ruined. There is currently no mechanism to detect and halt generation early when the model begins to degenerate.

### Root cause

The AR talker (`Qwen3TTSTalkerForConditionalGeneration`) generates codec tokens autoregressively. Each token's embedding feeds into the next step, so a few bad samples compound — the model loses coherence, entropy spikes (or collapses), and the output devolves into noise. The existing `repetition_penalty` (1.05) is too weak to prevent it, and `max_tokens` is too blunt — it's a ceiling, not a quality guard.

### Existing hooks (unused)

There is already an `_eos_logit_bias` field on the talker (`qwen3_tts_talker.py:330`) initialized to `0.0` and never modified. The `compute_logits` method (`qwen3_tts_talker.py:429-432`) applies it but since it's always zero, it's a no-op. This was clearly intended as a hook for exactly this kind of intervention.

## Proposed solution: entropy guardrail

Add an **optional, per-request entropy-based degeneration detector** that monitors the softmax entropy of the talker's logits at each decode step. When entropy exits a healthy range for several consecutive steps, the guardrail forces EOS and returns an error to the client.

### How it works

1. **Entropy computation** — After `compute_logits` applies the codec mask, compute Shannon entropy over the valid logit distribution:
   ```python
   probs = F.softmax(logits, dim=-1)
   entropy = -torch.sum(probs * torch.log(probs + 1e-10), dim=-1)  # per batch element
   ```

2. **Degeneration detection** — Track a per-request rolling window of entropy values. Flag degeneration when:
   - **Entropy spike**: entropy exceeds `entropy_threshold_high` for `N` consecutive steps (model is lost/uncertain)
   - **Entropy collapse**: entropy drops below `entropy_threshold_low` for `N` consecutive steps (model is stuck in a loop)
   
3. **Intervention** — When degeneration is detected:
   - Force EOS by setting the EOS logit to a very high value (e.g., `+100.0`) and all others to `-inf`
   - The vLLM sampler picks EOS, the AR scheduler sees the stop token, and the request finishes normally via `RequestStatus.FINISHED_STOPPED`
   - Set a flag in the request's `additional_information` (e.g., `entropy_guardrail_triggered: true`) so downstream stages and the API can detect it

4. **Error surfacing** — The serving layer checks for the flag and returns an appropriate error:
   - **Non-streaming**: HTTP 500 with message `"Speech generation halted: audio quality degeneration detected (entropy guardrail triggered)"`
   - **Streaming**: Log warning; the truncated audio is still usable up to the degeneration point (graceful degradation)

### API parameters

Add to `OpenAICreateSpeechRequest` (protocol/audio.py):

```python
entropy_guardrail: bool | None = Field(
    default=None,
    description="Enable entropy-based degeneration detection. When triggered, "
                "generation stops early to prevent noise output. Default: server config.",
)
entropy_threshold_high: float | None = Field(
    default=None, ge=0.0,
    description="Upper entropy threshold. Consecutive steps above this trigger halt. "
                "Default: 4.5 (log2 scale over ~2048 valid codec tokens).",
)
entropy_threshold_low: float | None = Field(
    default=None, ge=0.0,
    description="Lower entropy threshold. Consecutive steps below this trigger halt. "
                "Default: 0.5.",
)
entropy_window: int | None = Field(
    default=None, ge=1,
    description="Number of consecutive out-of-range steps before triggering. Default: 5.",
)
```

These flow through `_build_tts_params()` → `additional_information` → talker `preprocess()` → stored in `model_intermediate_buffer` per request.

#### 1. Protocol layer (`vllm_omni/entrypoints/openai/protocol/audio.py`)
- Add `entropy_guardrail`, `entropy_threshold_high`, `entropy_threshold_low`, `entropy_window` fields to `OpenAICreateSpeechRequest`

#### 2. Serving layer (`vllm_omni/entrypoints/openai/serving_speech.py`)
- In `_build_tts_params()`: pass entropy guardrail parameters through to `additional_information`
- In `_validate_tts_request()`: validate parameter ranges
- In `create_speech()` (non-streaming path): check for `entropy_guardrail_triggered` flag on final output and return error response
- In `_generate_pcm_chunks()` (streaming path): log warning when guardrail fires, let truncated audio through

#### 3. Talker preprocess (`vllm_omni/model_executor/models/qwen3_tts/qwen3_tts_talker.py`)
- In `preprocess()`: extract entropy guardrail config from `info_dict`, store in `model_intermediate_buffer` per request
- Add per-request state: `entropy_history` (deque), `entropy_guardrail_config` (thresholds + window), `entropy_guardrail_triggered` (bool)

#### 4. Talker compute_logits (`qwen3_tts_talker.py:415-434`)
- After codec mask application, compute entropy
- Look up per-request entropy state from `sampling_metadata.output_token_ids` (step count) and the model intermediate buffer
- **Challenge**: `compute_logits` operates on a batch and doesn't have direct access to `model_intermediate_buffer`. Two approaches:
  - **(a) Stateful class member**: Store entropy tracking state on `self` keyed by batch index, updated each step. The `preprocess`/`postprocess` hooks manage lifecycle.
  - **(b) Via sampling_metadata**: Access `sampling_metadata.output_token_ids[i]` for step count, maintain a dict keyed by tuple of recent token ids.
  - **Recommended: (a)** — preprocess already manages per-request state; add entropy tracking there.

#### 5. Talker postprocess (`qwen3_tts_talker.py:630-635`)
- If guardrail triggered, set `entropy_guardrail_triggered: true` in the returned info dict
- This propagates through `model_intermediate_buffer` → `make_omni_output` → downstream

#### 6. Stage config (`vllm_omni/model_executor/stage_configs/qwen3_tts.yaml`)
- Add server-wide defaults under `default_sampling_params` or a new `tts_args` section:
  ```yaml
  tts_args:
    entropy_guardrail: true
    entropy_threshold_high: 4.5
    entropy_threshold_low: 0.5
    entropy_window: 5
  ```

### Design considerations

- **Batch-aware**: The guardrail must work correctly when multiple requests are batched. Entropy state is per-request, not per-batch.
- **Minimal overhead**: Entropy is a single `softmax + sum` over the logits tensor that's already computed. The cost is negligible compared to the transformer forward pass.
- **No false positives on natural variation**: The `entropy_window` (consecutive steps) prevents triggering on momentary entropy fluctuations that are normal in speech.
- **Graceful for streaming**: In streaming mode, the client has already received valid audio up to the degeneration point. Halting early is better than continuing with noise.
- **Opt-out**: Can be disabled per-request with `entropy_guardrail: false`.

### Calibration needed

Before shipping, we need to empirically determine good default thresholds:
- Generate a batch of healthy TTS outputs and log per-step entropy to establish the normal range
- Generate known-bad outputs (by seeding with adversarial inputs or low repetition_penalty) and observe entropy patterns
- The defaults in this issue (high=4.5, low=0.5, window=5) are educated guesses and should be validated

### Files to modify

| File | Change |
|------|--------|
| `vllm_omni/entrypoints/openai/protocol/audio.py` | Add request fields |
| `vllm_omni/entrypoints/openai/serving_speech.py` | Pass params, check trigger flag, surface error |
| `vllm_omni/model_executor/models/qwen3_tts/qwen3_tts_talker.py` | Entropy computation, tracking, EOS forcing |
| `vllm_omni/model_executor/stage_configs/qwen3_tts.yaml` | Server-wide defaults |
| `vllm_omni/model_executor/stage_configs/qwen3_tts_batch.yaml` | Same defaults |
| `vllm_omni/model_executor/stage_configs/qwen3_tts_no_async_chunk.yaml` | Same defaults |
